Make c10::util::get_fully_qualified_type_name() backwards compatible with clang 4

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26482 [wip] Error checking for KernelFunction::callUnboxed()
* #30712 Better error messages thanks to compile time type names
* #31307 [no-land] try fix 2
* #31306 [no-land] try fix 1
* #26628 Remove CAFFE_KNOWN_TYPE
* #31313 Fix c10::util::get_fully_qualified_type_name for MSVC
* **#31351 Make c10::util::get_fully_qualified_type_name() backwards compatible with clang 4**
* #31312 constexpr typeid improvements

Clang 4 needs the c10:: namespace specifier on fully_qualified_type_name_impl() to work correctly.

Also, let's add an error message for people using clang 3 and earlier, we don't support those compilers anymore but before this PR, they got a crappy message.

Differential Revision: [D19135587](https://our.internmc.facebook.com/intern/diff/D19135587/)

smessmer committed:
Make c10::util::get_fully_qualified_type_name() backwards compatible with clang 4

Clang 4 needs the c10:: namespace specifier on fully_qualified_type_name_impl() to work correctly.

Also, let's add an error message for people using clang 3 and earlier, we don't support those compilers anymore but before this PR, they got a crappy message.

Differential Revision: [D19135587](https://our.internmc.facebook.com/intern/diff/D19135587/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit f9ceb69:
* **1/1** broken upstream at merge base 643ca5d (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=643ca5def2ea140f16bb064740b8de331ee465d8&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f9ceb698ca1551ced0ba1c0e0e63ee158b550f77).

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py2_gcc5_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75766088)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

dzhulgakov reviewed (approved):
Looks good!

smessmer committed:
Update on "Make c10::util::get_fully_qualified_type_name() backwards compatible with clang 4"


Clang 4 needs the c10:: namespace specifier on fully_qualified_type_name_impl() to work correctly.

Also, let's add an error message for people using clang 3 and earlier, we don't support those compilers anymore but before this PR, they got a crappy message.

Differential Revision: [D19135587](https://our.internmc.facebook.com/intern/diff/D19135587/)

[ghstack-poisoned]

smessmer committed:
Update on "Make c10::util::get_fully_qualified_type_name() backwards compatible with clang 4"


Clang 4 needs the c10:: namespace specifier on fully_qualified_type_name_impl() to work correctly.

Also, let's add an error message for people using clang 3 and earlier, we don't support those compilers anymore but before this PR, they got a crappy message.

Differential Revision: [D19135587](https://our.internmc.facebook.com/intern/diff/D19135587/)

[ghstack-poisoned]
